t/unit/tasks/test_canvas.py

```diff
@@ -816,21 +833,21 @@ def test_repr(self):
         assert 'without body' in repr(x)
 
     def test_freeze_tasks_body_is_group(self):
-        # Confirm that `group index` is passed from a chord to elements of its
-        # body when the chord itself is encapsulated in a group
+        # Confirm that `group index` values counting up from 0 are set for
+        # elements of a chord's body when the chord is encapsulated in a group
         body_elem = self.add.s()
-        chord_body = group([body_elem])
+        chord_body = group([body_elem] * 42)
         chord_obj = chord(self.add.s(), body=chord_body)
         top_group = group([chord_obj])
         # We expect the body to be the signature we passed in before we freeze
-        (embedded_body_elem, ) = chord_obj.body.tasks
-        assert embedded_body_elem is body_elem
-        assert embedded_body_elem.options == dict()
-        # When we freeze the chord, its body will be clones and options set
+        for embedded_body_elem in chord_obj.body.tasks:
+            assert embedded_body_elem is body_elem
+            assert not embedded_body_elem.options
+        # When we freeze the chord, its body will be cloned and options set
         top_group.freeze()
-        (embedded_body_elem, ) = chord_obj.body.tasks
-        assert embedded_body_elem is not body_elem
-        assert embedded_body_elem.options["group_index"] == 0   # 0th task
+        for i, embedded_body_elem in enumerate(chord_obj.body.tasks):
+            assert embedded_body_elem is not body_elem
+            assert embedded_body_elem.options["group_index"] == i
```

Comment:
I think that this loop for example is a great opportunity to introduce [`pytest-subtests`](https://github.com/pytest-dev/pytest-subtests/) to our codebase.

There are other opportunities in this patch to use it.

Reply:
I've pivoted to using `all(...)` for the other loopy checks in this diff now, so this should be the only place subtests might be relevant. I'll do it here to support the first commit and introduce the subtest tooling so that it can be used elsewhere in the suite.

Reply:
Oh, I actually see what you mean. We could use subtests to check, e.g., that the body of a deeply deserialized chord is correct, separate from the checks on the header! That's pretty cool, I'll do that too.

Reply:
Done, check out the new diff and let me know what you think